opengrok-indexer/src/main/java/org/opengrok/indexer/util/Progress.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2007, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Portions Copyright (c) 2017-2018, Chris Fraire <cfraire@me.com>.
+ */
+
+package org.opengrok.indexer.util;
+
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class Progress implements AutoCloseable {
+    private final Logger logger;
+    private final long totalCount;
+    private final String suffix;
+
+    private AtomicLong currentCount = new AtomicLong();
+    private Thread loggerThread = null;
+    private volatile boolean run;
+
+    private final Object sync = new Object();
+
+    /**
+     * @param logger logger instance
+     * @param suffix string suffix to identify the operation
+     * @param totalCount total count
+     */
+    public Progress(Logger logger, String suffix, long totalCount) {
+        this.logger = logger;
+        this.suffix = suffix;
+        this.totalCount = totalCount;
+
+        // Assuming printProgress configuration setting cannot be changed on the fly.
+        if (totalCount > 0 && RuntimeEnvironment.getInstance().isPrintProgress()) {
+            // spawn a logger thread.
+            run = true;
+            loggerThread = new Thread(this::logLoop,
+                    "progress-thread-" + suffix.replaceAll(" ", "_"));
+            loggerThread.start();
+        }
+    }
+
+    /**
+     * Increment counter. The actual logging will be done eventually.
+     */
+    public void increment() {
+        this.currentCount.incrementAndGet();
+
+        if (loggerThread != null) {
+            // nag the thread.
+            synchronized (sync) {
+                sync.notify();
+            }
+        }
+    }
+
+    private void logLoop() {
+        long cachedCount = 0;
+
+        while (true) {
+            if (!run) {
+                return;
+            }
+
+            long currentCount = this.currentCount.get();
+            Level currentLevel;
+
+            // Do not log if there was no progress.
+            if (cachedCount >= currentCount) {
+                continue;
```

Comment:
[Spurious wakeups](http://tutorials.jenkov.com/java-concurrency/thread-signaling.html#spurious-wakeups) are a thing. If so, and you get here, then you have a tight loop with a `continue` burning CPU. I recommend just making the logging below conditional on `currentCount > cachedCount` and proceeding to `wait` again.

Reply:
I actually encountered the active wait myself when adding a test.